context/http.go

```diff
@@ -197,13 +195,7 @@ func (ctx *httpRequestContext) Value(key interface{}) interface{} {
 			goto fallback
 		}
 
-		parts := strings.Split(keyStr, ".")
-
-		if len(parts) != 3 {
-			goto fallback
-		}
-
-		switch parts[2] {
+		switch strings.TrimPrefix(keyStr, "http.request.") {
```

Comment:
Since we already know (from line 194) that `keyStr` has `"http.request."` as its prefix, unconditionally slicing the string by a constant would be more optimal.
```suggestion
		switch keyStr[len("http.request."):] {
```
[The compiler is smart enough to evaluate `len(constant)` at compile time.](https://go.godbolt.org/z/6P91svb5a)

Reply:
the thing about slicing the `string`s like this always makes me feel uneasy because Go slices `string`s by `bytes` not by `rune`s, so I'm happy with `Trim`ming in this case because it avoids "bad habits"

Reply:
Yes, I must admit I went for "readability" here.

Reply:
Actually, we can completely drop the `strings.HasPrefix()`. `strings.TrimPrefix()` already handles that, and if there's no match in the switch, we'll continue with the fallback.

Changing that also makes the code easier to read, and we don't have to use the ugly `goto fallback`


```patch
diff --git a/context/http.go b/context/http.go
index 7471e572..7fb75790 100644
--- a/context/http.go
+++ b/context/http.go
@@ -191,10 +191,6 @@ func (ctx *httpRequestContext) Value(key interface{}) interface{} {
 			return ctx.r
 		}
 
-		if !strings.HasPrefix(keyStr, "http.request.") {
-			goto fallback
-		}
-
 		switch strings.TrimPrefix(keyStr, "http.request.") {
 		case "uri":
 			return ctx.r.RequestURI
@@ -220,10 +216,11 @@ func (ctx *httpRequestContext) Value(key interface{}) interface{} {
 			if ct != "" {
 				return ct
 			}
+		default:
+			// no match; fall back to standard behavior below
 		}
 	}
 
-fallback:
 	return ctx.Context.Value(key)
 }
 
@@ -296,10 +293,6 @@ func (irw *instrumentedResponseWriter) Value(key interface{}) interface{} {
 			return irw
 		}
 
-		if !strings.HasPrefix(keyStr, "http.response.") {
-			goto fallback
-		}
-
 		irw.mu.Lock()
 		defer irw.mu.Unlock()
 
@@ -313,9 +306,10 @@ func (irw *instrumentedResponseWriter) Value(key interface{}) interface{} {
 			if contentType != "" {
 				return contentType
 			}
+		default:
+			// no match; fall back to standard behavior below
 		}
 	}
 
-fallback:
 	return irw.Context.Value(key)
 }
```

Reply:
> Actually, we can completely drop the `strings.HasPrefix()`. `strings.TrimPrefix()` already handles that, and if there's no match in the switch, we'll continue with the fallback.

Not really: `strings.TrimPrefix` returns the input string unchanged if there is no match, so the switch would match both `keyStr == "uri"` and `"http.request.uri"` if we drop the `strings.HasPrefix()` check.

Reply:
🤦 yup

Reply:
Updated to use the full string in the switch (as we discussed) 👍